src/operations/blas3/gemm_common.hpp

```diff
@@ -42,13 +42,22 @@ mul_add(T a, T b, T c,
 }
 #endif
 
-template <typename T>
-static PORTBLAS_INLINE T
-mul_add(T a, T b, T c,
-        typename std::enable_if<is_sycl_scalar<T>::value>::type * = 0) {
+template <typename Tin, typename Tout>
+static PORTBLAS_INLINE Tout
+mul_add(Tin a, Tin b, Tout c,
+        typename std::enable_if<std::is_same<Tin, Tout>::value &&
+                                is_sycl_scalar<Tin>::value>::type * = 0) {
   return (cl::sycl::mad(a, b, c));
 }
 
+template <typename Tin, typename Tout>
+static PORTBLAS_INLINE Tout
+mul_add(Tin a, Tin b, Tout c,
+        typename std::enable_if<!std::is_same<Tin, Tout>::value &&
+                                is_sycl_scalar<Tin>::value>::type * = 0) {
+  return static_cast<Tout>(a) * static_cast<Tout>(b) + c;
```

Comment:
@hjabird yes indeed, it should be actually done in `Tin` but I had to fall back to `Tout` because gemm unit-tests failed on the oneMKL side (portBLAS backend) due to precision mismatches. This needs to be investigated though I agree, as casting to float beats the purpose of having half mul.

Reply:
Do we have any ticket or issue to track that this needs to be investigated?

Reply:
@hjabird not that I know of no, I've put a placeholder [oneMKL PR](https://github.com/OuadiElfarouki/oneMKL/pull/1) locally that we can open once the portBLAS PRs are merged. Something similar was raised in this [PR](https://github.com/oneapi-src/oneMKL/pull/466/files#r1552523436) regarding the test precision tolerance. So basically if the precision on oneMKL tests is handled loosely for half/mixed operations, we can drop this static_cast on portBLAS side as it only causes failures on oneMKL tests not on our portBLAS unittests.